File: Lucy/LeetCode 75/Array_String/1071_Greatest_Common_Divisor_of_Strings.js
/**
 * 1071. Greatest Common Divisor of Strings
 *
 * url: https://leetcode.com/problems/greatest-common-divisor-of-strings/?envType=study-plan-v2&envId=leetcode-75
 * difficulty: Easy
 * topic:
 * date: 2024.10.11(FRI)
 */

/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */
var gcdOfStrings = function (str1, str2) {
  let answer = "";

  const isDivided = (str, divisor) =>
    str.split(divisor).filter((s) => s.length !== 0).length === 0;

  const shorter = str1.length > str2.length ? str2 : str1;

  for (let i = 0; i < shorter.length; i++) {
    const divisor = shorter.slice(0, i + 1);
    if (isDivided(str1, divisor) && isDivided(str2, divisor)) answer = divisor;
  }
  return answer;
};

console.log(gcdOfStrings("ABCABC", "ABC")); // ABC
console.log(gcdOfStrings("ABABAB", "ABAB")); // AB
console.log(gcdOfStrings("LEET", "CODE")); // ""
console.log(gcdOfStrings("ABABABAB", "ABAB")); // "ABAB"

File: Seori/LeetCode 75/Array_String/1071. Greatest Common Divisor of Strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        # [1] 보다 짧은 문자열을 찾는다.
        if len(str1) > len(str2):
            longer, shorter = str1, str2
        else:
            longer, shorter = str2, str1

        # [2] [1]에서 찾아낸 짧은 문자열을 앞에서부터 [0:i+1]만큼 잘라서 divisor라고 명명한다.
        answer = ''
        for i in range(len(shorter)):
            divisor = shorter[0:i+1]

            # [3] divisor가 짧은 문자열을 나눌 수 있고, 긴 문자열 또한 나눌 수 있다면 answer로 갱신한다.
            if divisor * (len(shorter) // len(divisor)) == shorter:
                if divisor * (len(longer) // len(divisor)) == longer:
                    answer = divisor
        
        return answer